DIPs/DIP27.md

```diff
@@ -0,0 +1,172 @@
+
+## Description
+
+Function are still defined with the same syntax :
+
+```d
+ReturnType functionName(Parameters) {
+  // Function body.
+}
+```
+
+However, this is now strictly equivalent to :
+
+```d
+enum functionName = function ReturnType(Parameters) {
+  // Function body.
+};
+```
+
+The function has a function type and can be used as this anywhere a function is expected :
+
+```d
+static assert(is(typeof(functionName) == ReturnType function(Parameters)); // OK
+
+ReturnType function(Parameters) foo = functionName; // OK
+
+void buzz(ReturnType function(Parameters) qux);
+buzz(foo); // OK
+buzz(bar); // OK
+
+auto a = functionName;
+static assert(is(typeof(a) == ReturnType function(Parameters)); // OK
+
+auto b = functionName();
+static assert(is(typeof(b) == ReturnType); // OK
+```
+
+Note that what is above is simply consequences of the simple rule expressed above.
+
+### Is Expression
+
+The function type specialization used in is expression has to be modified as well. It does match the first class function we defined above, and alias parameters as defined on https://dlang.org/spec/expression.html#IsExpression .
+
+```d
+void foo() {}
+auto bar = foo;
+
+static assert(is(typeof(foo) == function)); // OK
+static assert(is(typeof(bar) == function)); // OK
+static assert(is(void function() == function)); // OK
+```
+
+Matching parameters using an identifier is also affected to work with all first class functions.
+
+```d
+static assert(is(typeof(foo) P == function)); // OK, P is an empty sequence.
+static assert(is(void function(uint) P == function)); // OK, P is an a sequence of one element : uint.
+```
+
+Alterantively, this behavior may be dropped as '''__parameters''' already provides it.
+
+### Optional parentheses
+
+Redundant parenthesis can be a burden for argument-less function class (or single argument function calls used as UFCS). An implicit function call is performed in the following cases :
+
+1. When .identifier lookup fails on the function :
+
+```d
+uint foo() {
+  return 0;
+}
+
+void bar(uint function() a) { writeln("function bar"); }
+void bar(uint a) { writeln("uint bar"); }
+
+foo.bar(); // function bar
+foo().bar(); // uint bar
+
+void buzz(uint a) {}
+foo.buzz(); // OK: Implicit call to foo is added as lookup failed on the function.
+```
+
+2. When used in a foreach :
+
+```d
+import std.algorithm;
+ 
+void main() {
+  foreach(i; iota(5).map!(n => n * n)) {
+    import std.stdio;
+    writeln(i); // Prints 1 then 4, 9, 16 and 25.
+  }
+}
+```
+
+3. When used as an expression statement that has no effect :
+
+```d
+uint foo() {
+  return 0;
+}
+
+auto bar = foo;
+
+foo; // Will call foo.
+bar; // Will call foo as well.
+```
+
+You must note that functions are regular first class function. So optional parentheses apply regardless of how is defined the function :
```

Comment:
I think it should be "regardless how the function is defined" or "regardless of the function's definition"